shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -47,71 +68,116 @@ public String getCommandName() {
-      runWildCardCmd(inputPath, cl);
+      candidateUris.addAll(FileSystemShellUtils.getAlluxioURIs(mFileSystem, new AlluxioURI(path)));
+    }
+    final BlockingQueue<AlluxioURI> toPersist = new LinkedBlockingQueue<>();
+    for (AlluxioURI uri : candidateUris) {
+      queuePersist(mFileSystem.getStatus(uri), toPersist);
+    }
+    System.out.println("Found " + toPersist.size() + " files to persist.");
+    ExecutorService service =
+        Executors.newFixedThreadPool(parallelism, ThreadFactoryUtils.build("persist-cli-%d", true));
+    final Object progressLock = new Object();
+    AtomicInteger completedFiles = new AtomicInteger(0);
+    List<Future<Void>> futures = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      futures.add(service.submit(new PersistCallable(toPersist, completedFiles, progressLock)));
+    }
+    try {
+      for (Future<Void> future : futures) {
+        future.get();
+      }
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+      return -1;
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
+      return -1;
+    } finally {
+      service.shutdownNow();
     }
     return 0;
   }
 
+  private void queuePersist(URIStatus status, BlockingQueue<AlluxioURI> toPersist)
+      throws AlluxioException, IOException {
+    AlluxioURI uri = new AlluxioURI(status.getPath());
+    if (status.isFolder()) {
+      List<URIStatus> statuses = mFileSystem.listStatus(uri);
+      for (URIStatus s : statuses) {
+        queuePersist(s, toPersist);
+      }
+    } else if (!status.isPersisted()) {
+      toPersist.add(uri);
+    }
+  }
+
   /**
-   * Persists a file or directory currently stored only in Alluxio to the UnderFileSystem.
-   *
-   * @param filePath the {@link AlluxioURI} path to persist to the UnderFileSystem
+   * Thread that polls a persist queue and persists files.
    */
-  private void persist(AlluxioURI filePath) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(filePath);
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
+  private class PersistCallable implements Callable<Void> {
+    private final BlockingQueue<AlluxioURI> mToPersist;
+    private final int mTotalFiles;
+    private final Object mProgressLock;
+    private final AtomicInteger mCompletedFiles;
+
+    PersistCallable(BlockingQueue<AlluxioURI> toPersist, AtomicInteger count, Object progressLock) {
+      mToPersist = toPersist;
+      mTotalFiles = toPersist.size();
+      mProgressLock = progressLock;
+      mCompletedFiles = count;
+    }
+
+    @Override
+    public Void call() throws Exception {
+      AlluxioURI toPersist = mToPersist.poll();
```

Comment:
nit: we called the queue `toPersist` in the constructor, can we call this something else, e.g. `nextFile`?